Mathlib/Topology/Order/NhdsSet.lean

```diff
@@ -0,0 +1,161 @@
+/-
+Copyright (c) 2023 Yury Kudryashov. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Yury Kudryashov
+-/
+import Mathlib.Topology.Order.Basic
+
+/-!
+# Set neighborhoods of intervals
+
+In this file we prove basic theorems about `𝓝ˢ s`,
+where `s` is one of the intervals
+`Set.Ici`, `Set.Iic`, `Set.Ioi`, `Set.Iio`, `Set.Ico`, `Set.Ioc`, `Set.Ioo`, and `Set.Icc`.
+
+First, we prove lemmas in terms of filter equalities,
+then we prove lemmas about `s ∈ 𝓝ˢ t`, where both `s` and `t` are intervals.
+-/
+
+
+open Set Filter OrderDual
+open scoped Topology
+
+variable {α : Type _} [LinearOrder α] [TopologicalSpace α] [OrderClosedTopology α] {a b c d : α}
+
+/-!
+# Formulae for `𝓝ˢ` of intervals
+-/
+
+@[simp] theorem nhdsSet_Ioi : 𝓝ˢ (Ioi a) = 𝓟 (Ioi a) := isOpen_Ioi.nhdsSet_eq
+@[simp] theorem nhdsSet_Iio : 𝓝ˢ (Iio a) = 𝓟 (Iio a) := isOpen_Iio.nhdsSet_eq
+@[simp] theorem nhdsSet_Ioo : 𝓝ˢ (Ioo a b) = 𝓟 (Ioo a b) := isOpen_Ioo.nhdsSet_eq
+
+theorem nhdsSet_Ici : 𝓝ˢ (Ici a) = 𝓝 a ⊔ 𝓟 (Ioi a) := by
+  rw [← Ioi_insert, nhdsSet_insert, nhdsSet_Ioi]
+
+theorem nhdsSet_Iic : 𝓝ˢ (Iic a) = 𝓝 a ⊔ 𝓟 (Iio a) := nhdsSet_Ici (α := αᵒᵈ)
+
+theorem nhdsSet_Ico (h : a < b) : 𝓝ˢ (Ico a b) = 𝓝 a ⊔ 𝓟 (Ioo a b) := by
+  rw [← Ioo_insert_left h, nhdsSet_insert, nhdsSet_Ioo]
+
+theorem nhdsSet_Ioc (h : a < b) : 𝓝ˢ (Ioc a b) = 𝓝 b ⊔ 𝓟 (Ioo a b) := by
+  rw [← Ioo_insert_right h, nhdsSet_insert, nhdsSet_Ioo]
+
+theorem nhdsSet_Icc (h : a ≤ b) : 𝓝ˢ (Icc a b) = 𝓝 a ⊔ 𝓝 b ⊔ 𝓟 (Ioo a b) := by
+  rcases h.eq_or_lt with rfl | hlt
+  · simp
+  · rw [← Ioc_insert_left h, nhdsSet_insert, nhdsSet_Ioc hlt, sup_assoc]
+
+/-!
+### Lemmas about `Ixi _ ∈ 𝓝ˢ (Set.Ici _)`
+-/
+
+@[simp]
+theorem Ioi_mem_nhdsSet_Ici : Ioi a ∈ 𝓝ˢ (Ici b) ↔ a < b := by
+  refine ⟨(Ici_subset_Ioi.1 <| principal_le_nhdsSet ·), fun h ↦ ?_⟩
+  rw [nhdsSet_Ici]
+  exact ⟨Ioi_mem_nhds h, Ioi_subset_Ioi h.le⟩
+
+theorem Ici_mem_nhdsSet_Ici (h : a < b) : Ici a ∈ 𝓝ˢ (Ici b) :=
+  mem_of_superset (Ioi_mem_nhdsSet_Ici.2 h) Ioi_subset_Ici_self
```

Comment:
This would be an `iff` in the order topology, right? I don't want you to add it, but that makes me think that the naming of these could get quickly inconsistent. I can think of two options:
- add `_iff` to the iff lemmas, and add aliases for the backwards directions with the current names
- add `of_lt` to the ones which aren't iff, and add aliases for backwards directions of `iff`s with `of_lt`

Reply:
I added `_iff` to the `iff` lemmas. The `Ici_mem_nhdsSet_Ici` is not an `iff` in the order topology, e.g., for `a = b = 0` in `Nat`.

Reply:
Oh right we would probably need `NoBotOrder` for it to work.

Reply:
Oh actually you proved it at the end of the file, perfect